### DISCUSSION TITLE
Is it possible change quartz.dataSource.myDS.connectionString when the apps is running?

### DISCUSSION THREAD

**Discussion by guillermo2007** [quartznet/quartznet#1863] [General]
Hello everyone

Is there any possibility to change the connection string from AdoJobStore in SqlServer when the application is running?

My idea is to have some databases in SqlServer and create a web application that can be administrate the jobs of each database.

For this, i need to change the instance to database when the user selected the other one

Any idea?

Thanks

**Comment by lahma** [comment]
You need to create a scheduler per connection string. You can configure it with `ZeroSizeThreadPool` and don't start it so it will be lightweight maintenance API.

**Reply by guillermo2007** [reply]
Thank you for replying.

How can i configure these? I can´t paste code by the privacity, but when i instance at startup file, i have written this: 

`services.AddQuartz(q =>
{
    // Use a Scoped container to create jobs.
    q.UseMicrosoftDependencyInjectionJobFactory();
});

// Add the Quartz.NET hosted service
services.AddQuartzHostedService(q => q.WaitForJobsToComplete = true);`

**Reply by lahma** [reply]
That's for server processing. Try using `SchedulerBuilder` to create a plain scheduler instance.

**Reply by guillermo2007** [reply]
Thanks Marko

is it neccesary register some instances from "QuartzOptions" at startup application?

Something like this:
 foreach(var instance in instances)
{
	services.Configure<QuartzOptions>(options =>
	{                
			options.Scheduling.IgnoreDuplicates = instance.IIgnoreDuplicates; // default: false
			options.Scheduling.OverWriteExistingData = true; // default: true                   
	});
}

**Reply by lahma** [reply]
For maintenance, you can create the needed non-running schedulers using SchedulerBuilder, per connection string. Then you need to have the actual worker schedulers configured they way you need, using whatever configuration system is the best fit for you.

**Reply by guillermo2007** [reply]
Perfect, I will do it like this

**Comment by guillermo2007** [comment]
Hello Marko,

I have a problem,

I have two pieces of code:

1. I create a custom Factory instance to IScheduler. i do this:

```c#
public async Task<IScheduler> GetSchedulerSelected()
{
    try
    {
        using var scope = serviceScopeFactory.CreateScope();
        var coreSchedulerDbContext = scope.ServiceProvider.GetRequiredService<CoreSchedulerDbContext>();

        var tenant = await coreSchedulerDbContext.Tenants.FirstOrDefaultAsync(x => x.Selected);

        if (tenant == null )
            throw new TenantNotFoundException();                

        if (string.IsNullOrEmpty(tenant.ConnectionString))
            throw new TenantHasNotConnectionStringException();

        var properties = new NameValueCollection
        {
            ["quartz.scheduler.instanceName"] = tenant?.Name,
            ["quartz.jobStore.type"] = "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz",
            ["quartz.jobStore.driverDelegateType"] = "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz",
            ["quartz.jobStore.useProperties"] = "false",
            ["quartz.jobStore.performSchemaValidation"] = "false",
            // "json" is alias for "Quartz.Simpl.JsonObjectSerializer, Quartz.Serialization.Json" 
            ["quartz.serializer.type"] = "json",
            ["quartz.jobStore.dataSource"] = "default",
            ["quartz.dataSource.default.connectionString"] = tenant.ConnectionString,
            ["quartz.dataSource.default.provider"] = "SqlServer"
        };
        ISchedulerFactory schdFactory = new StdSchedulerFactory(properties);                
        return await schdFactory.GetScheduler();
    }
    catch (Exception)
    {
        throw;
    }
}
``` 

2.-The next step is to get JobGroupNames and it doesn't return nothing, on the other hand, when i instance IScheduler by ISchedulerFactory, the methods return some data.

```c#
public async IAsyncEnumerable<ScheduleModel> GetAllJobsAsync(ScheduleJobFilter? filter = null)
{

    var schedulerCustom = await scheduleCustomFactory.GetSchedulerSelected();            

    var jobGroupNames = await schedulerCustom.GetJobGroupNames();

    foreach (var jobGrp in jobGroupNames)
    {
        if (filter != null && !filter.IncludeSystemJobs)
        {
            if (jobGrp == Constants.SYSTEM_GROUP)
                continue;
        }

        var jobKeys = await schedulerCustom.GetJobKeys(GroupMatcher<JobKey>.GroupEquals(jobGrp));

        foreach (var jobKey in jobKeys)
        {
            await foreach (var job in GetScheduleModelsAsync(jobKey))
            {
                yield return job;
            }
        }
    }
}
```

Do you know why?

Thanks

**Reply by lahma** [reply]
No idea, maybe check the database and make sure data is there and you have correct scheduler names configured, selecting data uses scheduler as filter criteria.

**Reply by guillermo2007** [reply]
Ok,

The data are in their databases, maybe de name of Scheduler, i will check. Is it the scheduler the only criteria or there are more?

Thanks

**Reply by lahma** [reply]
Should be the only one if I remember correctly.

**Reply by guillermo2007** [reply]
Ok, I have already checked and it's correct.

Thanks Marko
